polygeist/lib/Dialect/Polygeist/Transforms/ArgumentPromotion.cpp

```diff
@@ -475,28 +527,64 @@ bool ArgumentPromotionPass::isCandidateCallable(
   ModuleOp module = getOperation();
   SymbolTableCollection symTable;
   SymbolUserMap userMap(symTable, module);
-  [[maybe_unused]] StringRef calleeName =
+  [[maybe_unused]] StringRef callableName =
       cast<SymbolOpInterface>(callableOp.getOperation()).getName();
 
-  // The callee must be defined and private.
+  // The function must be defined and private.
   if (!isPrivateDefinition(callableOp)) {
-    LLVM_DEBUG({
-      llvm::dbgs() << "Not a candidate: " << calleeName << "\n";
-      llvm::dbgs().indent(2) << "callee not privately defined\n";
-    });
+    LLVM_DEBUG(llvm::dbgs().indent(2)
+               << "not a candidate: not privately defined\n");
     return false;
   }
 
-  // The callee must have a single call site.
-  unsigned numUsers = userMap.getUsers(callableOp).size();
-  if (numUsers != 1) {
-    LLVM_DEBUG({
-      llvm::dbgs() << "Not a candidate: " << calleeName << "\n";
-      llvm::dbgs().indent(2) << "has " << numUsers << " call sites\n";
-    });
+  // Functions with no arguments are not interesting.
+  if (callableOp.getCallableRegion()->args_empty()) {
+    LLVM_DEBUG(llvm::dbgs().indent(2) << "not a candidate: no arguments\n");
     return false;
   }
 
+  // Ensure all the call sites for this function are either in a GPU kernel or
+  // in a function that is called directly by a GPU kernel.
```

Comment:
Would it be better to restrict it by checking that the call site from the GPU kernel to the non GPU kernel function is itself a candidate, i.e., the call chain from the GPU kernel are all candidates.

Reply:
It would make the transformation less restrictive, but add some complexity. Added a TODO for now.